File: clase-5/usuarios/backend/src/controllers/mongodb/login.controller.ts
import {Request, Response} from "express";
import UserSchema from "../../models/mongodb/user.schema";
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";

const authLogin = async (req: Request, res: Response) => {

    const {email, password} = req.body;

    const user: any = await UserSchema.findOne({email: email}, {password: 1, email: 1, type: 1});

    if (!user) {
        return res.status(401).json({
            message: "Credenciales incorrectas",
            status: false
        })
    }

    const isMatch = await bcrypt.compare(password, user.password);
    if (!isMatch) {
        return res.status(401).json({
            message: "Credenciales incorrectas",
            status: false
        })
    }

    const token = jwt.sign(
        { id: user._id, email: user.email, type: user.type },
        process.env.JWT_SECRET || '',
        { expiresIn: '1d' }
    );

    return res.status(200).json({
        message: "Login exitoso",
        status: true,
        token: token
    })
}

export {authLogin}

File: clase-5/empleados/backend/app/controllers/login_controller.py
import os
from datetime import datetime, timedelta
from jose import jwt
from passlib.context import CryptContext
from app.models.user_model import UserModel
from app.schemas.user_schema import LoginRequest

# Configuración para el hash de contraseñas
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

class LoginController:
    def __init__(self):
        self.user_model = UserModel()

    async def login(self, login_data: LoginRequest):
        """
        Autentica a un usuario y genera un token JWT
        """
        # Buscar usuario por email
        user = await self.user_model.find_by_email(login_data.email)

        if not user:
            return {
                "message": "Credenciales incorrectas",
                "status": False
            }

        # Verificar contraseña
        if not self._verify_password(login_data.password, user["password"]):
            return {
                "message": "Credenciales incorrectas",
                "status": False
            }

        # Generar token JWT
        token = self._create_token({
            "id": str(user["_id"]),
            "email": user["email"],
            "type": user["type"]
        })

        return {
            "message": "Login exitoso",
            "status": True,
            "token": token
        }

    def _verify_password(self, plain_password: str, hashed_password: str) -> bool:
        """
        Verifica si la contraseña coincide con el hash almacenado
        """
        return pwd_context.verify(plain_password, hashed_password)

    def _create_token(self, data: dict) -> str:
        """
        Crea un token JWT con los datos proporcionados
        """
        to_encode = data.copy()

        # Establecer tiempo de expiración (1 día)
        expire = datetime.utcnow() + timedelta(days=1)
        to_encode.update({"exp": expire})

        # Crear token JWT
        encoded_jwt = jwt.encode(
            to_encode,
            os.getenv("JWT_SECRET", "your-secret-key"),
            algorithm="HS256"
        )

        return encoded_jwt
